nemo/collections/nlp/models/question_answering/qa_bert_model.py

```diff
@@ -0,0 +1,892 @@
+                end_indexes = self._get_best_indexes(end_logits[pos], n_best_size)
+                # if we could have irrelevant answers,
+                # get the min score of irrelevant
+                if version_2_with_negative:
+                    feature_null_score = start_logits[pos][0] + end_logits[pos][0]
+                    if feature_null_score < score_null:
+                        score_null = feature_null_score
+                        min_null_feature_index = feature_index
+                        null_start_logit = start_logits[pos][0]
+                        null_end_logit = end_logits[pos][0]
+                for start_index in start_indexes:
+                    for end_index in end_indexes:
+                        # We could hypothetically create invalid predictions,
+                        # e.g., predict that the start of the span is in the
+                        # question. We throw out all invalid predictions.
+                        if start_index >= len(feature.tokens):
+                            continue
+                        if end_index >= len(feature.tokens):
+                            continue
+                        if start_index not in feature.token_to_orig_map:
+                            continue
+                        if end_index not in feature.token_to_orig_map:
+                            continue
+                        if not feature.token_is_max_context.get(start_index, False):
+                            continue
+                        if end_index < start_index:
+                            continue
+                        length = end_index - start_index + 1
+                        if length > max_answer_length:
+                            continue
+                        prelim_predictions.append(
+                            _PrelimPrediction(
+                                feature_index=feature_index,
+                                start_index=start_index,
+                                end_index=end_index,
+                                start_logit=start_logits[pos][start_index],
+                                end_logit=end_logits[pos][end_index],
+                            )
+                        )
+
+            if version_2_with_negative:
+                prelim_predictions.append(
+                    _PrelimPrediction(
+                        feature_index=min_null_feature_index,
+                        start_index=0,
+                        end_index=0,
+                        start_logit=null_start_logit,
+                        end_logit=null_end_logit,
+                    )
+                )
+            prelim_predictions = sorted(prelim_predictions, key=lambda x: (x.start_logit + x.end_logit), reverse=True)
+
+            _NbestPrediction = collections.namedtuple("NbestPrediction", ["text", "start_logit", "end_logit"])
+
+            seen_predictions = {}
+            nbest = []
+            for pred in prelim_predictions:
```

Comment:
A lot of code here is quite unclear - I understand this was not something you wrote but if you have extra time, maybe breaking these up into methods will be helpful.

Reply:
especially lines 462 to 471 should be easily refactors into a method

Reply:
Another advantage to refactoring code that you didnt write yourself is that you get to understand code from the perspective of the reader.